### PR TITLE
utils/EGLUtils: allow generic types

### PR DESCRIPTION
**Pull request by lrusak**
This is a change that is a precursor to #18534.

We need to allow generic types to be used so we can support multiple windowing systems at the same time.

It will be interesting to see if android builds ok.

I'm also not sure if this is _functionally_ correct. Hopefully @yol can have a look.

### CODE REVIEW

**Comment by yol** [comment]
It's unfortunately not possible to get this 100% correct with base EGL, which is why EGL_EXT_platform_base was created in the first place. I guess depending on it is not an option?

Both relevant types (EGLNativeDisplayType and EGLNativeWindowType) are pointers *except* for EGLNativeWindowType on X11, which is unsigned long. So I guess this would fail on X11 platforms where sizeof(uintptr_t) != sizeof(unsigned long). Do we care?

File: xbmc/utils/EGLUtils.h [hunk 1]
```diff
@@ -14,6 +14,9 @@
 #include <string>
 #include <vector>
 
+// needed before including egl.h as it's defined in gbm.h implicitly
+// this allows us to support generic types using EGL_NO_X11
+#undef __GBM__
```
Comment (yol) [inline xbmc/utils/EGLUtils.h:19]: Can't really rely on this in a header, since `egl.h` might have already been included.

**Comment by yol** [comment]
Seems mesa also uses void* internally for every platform, so we should be fine I guess

File: xbmc/utils/EGLUtils.h (hunk 1 above)
Comment (lrusak) [reply xbmc/utils/EGLUtils.h:19]: Yea, I'm not really sure how to solve that :/

**Comment by lrusak** [comment]
> It's unfortunately not possible to get this 100% correct with base EGL, which is why EGL_EXT_platform_base was created in the first place. I guess depending on it is not an option?
> 

Yea I'm not sure, it seems our android implementation doesn't use `EGL_EXT_platform_base` and I'm not sure if it's able to or what android abi is required to use it.

> Both relevant types (EGLNativeDisplayType and EGLNativeWindowType) are pointers _except_ for EGLNativeWindowType on X11, which is unsigned long. So I guess this would fail on X11 platforms where sizeof(uintptr_t) != sizeof(unsigned long). Do we care?

I don't know? we can add an assert like we have for gbm?

**Comment by a1rwulf** [comment]
@yol Just for my understanding:
I had a conversation already with @lrusak and he tried following approach, ofc it didn't work out in real life - but from an conceptional view:

What if you just use `EGLNativeDisplayType` and nothing else?
Depending on what platform you build, it will resolve to the correct type.
If we define `NO_X11` it will resolve to `void*` for any platform, which would effectively be the same as using `void*` directly!?

I wonder where my mistake is :)

**Comment by yol** [comment]
> What if you just use EGLNativeDisplayType and nothing else?
> Depending on what platform you build, it will resolve to the correct type.

Well we are kind of building all platforms at once :D But EGLContext.cpp is only compiled once. So it really depends on what EGLNativeDisplayType resolves to. If all possible variants of it are compatible, there is no problem. Most of them are pointers anyway, so that is OK. In the end, the EGL library also can only implement one valid function signature.

> If we define EGL_NO_X11 it will resolve to void* for any platform, which would effectively be the same as using void* directly!?

Does it really? With `EGL_NO_X11`, it will resolve to void* for Display and uintptr_t for Pixmap. And only if you're not on GBM or Android.

**Comment by a1rwulf** [comment]
> > What if you just use EGLNativeDisplayType and nothing else?
> > Depending on what platform you build, it will resolve to the correct type.
> 
> Well we are kind of building all platforms at once :D But EGLContext.cpp is only compiled once. So it really depends on what EGLNativeDisplayType resolves to. If all possible variants of it are compatible, there is no problem. Most of them are pointers anyway, so that is OK. In the end, the EGL library also can only implement one valid function signature.
> 

Right, so if I look at the eglplatform header I understand 2 implications.
It's not designed to allow building X11 + non X11 in one binary.

`EGL_NO_X11` should be in line 97 so it's considered before any of the platform specific codepaths, which would reliably allow to build any platform besides X11.
See https://gitlab.freedesktop.org/mesa/mesa/-/blob/master/include/EGL/eglplatform.h#L125

But even then, it would mean that no userspace application nor library is supposed to use any type but:
```
EGLNativeDisplayType;
EGLNativePixmapType;
EGLNativeWindowType;
```

Stuff like https://github.com/NilsBrause/waylandpp/blob/master/include/wayland-egl.hpp#L42 prevents us from using only EGLNativeType typedefs cause of:
```
/home/lukas/Documents/git/xbmc/xbmc/windowing/wayland/WinSystemWaylandEGLContext.cpp:73:87: error: invalid conversion from ‘wl_egl_window*’ to ‘EGLNativeWindowType’ {aka ‘long unsigned int’} [-fpermissive]
   73 |   if (!m_eglContext.CreatePlatformSurface(m_nativeWindow.c_ptr(), m_nativeWindow.c_ptr()))
      |                                                                   ~~~~~~~~~~~~~~~~~~~~^~
      |                                                                                       |
      |                                                                                       wl_egl_window*
```

> > If we define EGL_NO_X11 it will resolve to void* for any platform, which would effectively be the same as using void* directly!?
> 
> Does it really? With `EGL_NO_X11`, it will resolve to void* for Display and uintptr_t for Pixmap. And only if you're not on GBM or Android.

This shouldn't be an issue given my above explanation makes any sense :D

Anyway, in the end it looks like we shouldn't really care about:
> So I guess this would fail on X11 platforms where sizeof(uintptr_t) != sizeof(unsigned long). Do we care?

Or put an assert in like we have in gbm.
Just wanted to understand if my assumptions about cross-platform development are correct or not.
Hope it makes somewhat sense.

**Comment by lrusak** [comment]
for gbm should I make a custom gbm header that unsets `__GBM__` so anywhere we include gbm.h we include that custom header instead? What is the best option here?

"system_gbm.h"
```
#include <gbm.h>
#undef __GBM__
```

**Comment by a1rwulf** [comment]
> for gbm should I make a custom gbm header that unsets `__GBM__` so anywhere we include gbm.h we include that custom header instead? What is the best option here?
> 
> "system_gbm.h"
> 
> ```
> #include <gbm.h>
> #undef __GBM__
> ```

mhm might be an option - but not entirely sure.
@yol any other ideas?

**Comment by yol** [comment]
> Right, so if I look at the eglplatform header I understand 2 implications.
> It's not designed to allow building X11 + non X11 in one binary.

Yes. But it goes even further. I reckon older EGL (i.e., before EGL_EXT_platform_base, which is part of EGL 1.5 - released 2014!) was never supposed to be used for more than one platform at the same time (no matter which ones), at all - neither on the library user nor on the library implementer side. So what mesa does - trying to guess which platform you want by dissecting the pointer you give to eglGetDisplay - is essentially a hack to work around this. You can take a look at it here: https://gitlab.freedesktop.org/mesa/mesa/-/blob/57d317865e7bee02a17efcde8beeb6a220f900f1/src/egl/main/egldisplay.c#L130-141 - it's quite atrocious IMO. All of this was solved with EGL_EXT_platform_base, but as @lrusak said most Android and other vendor blobs etc. probably don't support it and will never support it... In case you haven't seen it yet, the extension completely gets rid of the EGLNativeDisplayType etc. typedefs and just passes everything as void* in the API.

> EGL_NO_X11 should be in line 97 so it's considered before any of the platform specific codepaths, which would reliably allow to build any platform besides X11.

As said, the headers are not really supposed to support more than one platform (at the same time) anyway :) So the decision to put it on the same "priority level" as X11 makes sense in that regard. I think the define was added so that you could use EGL inside the X server because you could not include the X headers there due to some conflicts. They define some pretty common names like `Status`, `True` etc. which can screw up your compilation.

> But even then, it would mean that no userspace application nor library is supposed to use any type but:
> EGLNativeDisplayType;
> [...]

Yes, that is of course true for EGL without EGL_EXT_platform_base.

> So I guess this would fail on X11 platforms where sizeof(uintptr_t) != sizeof(unsigned long). Do we care?

Since mesa made the decision that casting any `EGLNativeWindowType` to void* is OK ( https://gitlab.freedesktop.org/mesa/mesa/-/blob/57d317865e7bee02a17efcde8beeb6a220f900f1/src/egl/main/eglapi.c#L996 ), I think we should just try and see what happens.

> "system_gbm.h"

An approach I started at one point was a system_egl.h header actually akin to our system_gl.h, but I never finished it. We would include it instead of the EGL headers everywhere and it would unset all the platform defines before including the EGL headers and then set them again if they were set before. So hopefully, we would always get the generic typedefs. What do you think about that?

**Comment by a1rwulf** [comment]
Thx for your explanation!
I like your idea.

**Comment by lrusak** [comment]
> > So I guess this would fail on X11 platforms where sizeof(uintptr_t) != sizeof(unsigned long). Do we care?
> 
> Since mesa made the decision that casting any `EGLNativeWindowType` to void* is OK ( https://gitlab.freedesktop.org/mesa/mesa/-/blob/57d317865e7bee02a17efcde8beeb6a220f900f1/src/egl/main/eglapi.c#L996 ), I think we should just try and see what happens.

My only concern is breaking android. I looked into using EGL_EXT_platform_base but it's only guaranteed in android 10 (api 29) so we can't move to that right now.

> > "system_gbm.h"
> 
> An approach I started at one point was a system_egl.h header actually akin to our system_gl.h, but I never finished it. We would include it instead of the EGL headers everywhere and it would unset all the platform defines before including the EGL headers and then set them again if they were set before. So hopefully, we would always get the generic typedefs. What do you think about that?

I'm fine with whatever. I tried to knock the usage of `<EGL/egl.h>` down in this PR as potential points for problem.

I'm just wanting to start moving on whatever we decide because it's know holding up #18534 which is likely to take a bit of time for review and would be nice to get in for the next alpha.

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
merge or are we doing something different?

**Comment by lrusak** [comment]
@yol I pushed a fixup, is that what you had in mind?

File: xbmc/utils/EGLUtils.h (hunk 1 above)
Comment (a1rwulf) [reply xbmc/utils/EGLUtils.h:19]: You'll want to remove this undef in your latest version, right?

File: xbmc/system_egl.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#undef __GBM__
+#define KODI__GBM__
+#endif
+
+#if defined(USE_X11)
+#undef USE_X11
+#define KODI_USE_X11
+#endif
+
+#define EGL_NO_X11
```
Comment (yol) [inline xbmc/system_egl.h:28]: Don't we have to define `MESA_EGL_NO_X11_HEADERS` too for older mesa?

**Review by yol — COMMENTED** [review]
Yeah I think that's pretty much the idea :+1:

File: xbmc/system_egl.h (hunk 2 above)
Comment (lrusak) [reply xbmc/system_egl.h:28]: updated thanks :)

File: xbmc/utils/EGLUtils.h (hunk 1 above)
Comment (lrusak) [reply xbmc/utils/EGLUtils.h]: removed thanks :)

**Review by yol — APPROVED** [review]

**Pull request merged**